platforms/core-execution/build-cache-example-client/src/main/java/org/gradle/caching/example/BuildCacheClientModule.java

```diff
@@ -0,0 +1,471 @@
+import org.gradle.caching.local.internal.LocalBuildCacheService;
+import org.gradle.caching.local.internal.TemporaryFileFactory;
+import org.gradle.internal.concurrent.DefaultExecutorFactory;
+import org.gradle.internal.concurrent.ExecutorFactory;
+import org.gradle.internal.file.FileAccessTimeJournal;
+import org.gradle.internal.file.FileAccessTracker;
+import org.gradle.internal.file.FileMetadataAccessor;
+import org.gradle.internal.file.StatStatistics;
+import org.gradle.internal.file.TreeType;
+import org.gradle.internal.file.impl.SingleDepthFileAccessTracker;
+import org.gradle.internal.file.nio.ModificationTimeFileAccessTimeJournal;
+import org.gradle.internal.file.nio.NioFileMetadataAccessor;
+import org.gradle.internal.file.nio.PosixJdk7FilePermissionHandler;
+import org.gradle.internal.hash.DefaultFileHasher;
+import org.gradle.internal.hash.DefaultStreamHasher;
+import org.gradle.internal.hash.FileHasher;
+import org.gradle.internal.hash.StreamHasher;
+import org.gradle.internal.operations.BuildOperationDescriptor;
+import org.gradle.internal.operations.BuildOperationIdFactory;
+import org.gradle.internal.operations.BuildOperationListener;
+import org.gradle.internal.operations.BuildOperationProgressEventEmitter;
+import org.gradle.internal.operations.BuildOperationRunner;
+import org.gradle.internal.operations.BuildOperationState;
+import org.gradle.internal.operations.BuildOperationTimeSupplier;
+import org.gradle.internal.operations.CurrentBuildOperationRef;
+import org.gradle.internal.operations.DefaultBuildOperationIdFactory;
+import org.gradle.internal.operations.DefaultBuildOperationProgressEventEmitter;
+import org.gradle.internal.operations.DefaultBuildOperationRunner;
+import org.gradle.internal.operations.OperationFinishEvent;
+import org.gradle.internal.operations.OperationIdentifier;
+import org.gradle.internal.operations.OperationProgressEvent;
+import org.gradle.internal.operations.OperationStartEvent;
+import org.gradle.internal.snapshot.SnapshotHierarchy;
+import org.gradle.internal.snapshot.impl.DirectorySnapshotterStatistics;
+import org.gradle.internal.time.TimestampSuppliers;
+import org.gradle.internal.vfs.FileSystemAccess;
+import org.gradle.internal.vfs.VirtualFileSystem;
+import org.gradle.internal.vfs.impl.DefaultFileSystemAccess;
+import org.gradle.internal.vfs.impl.DefaultSnapshotHierarchy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.nio.file.Files;
+import java.util.Collections;
+import java.util.function.Supplier;
+
+import static org.gradle.cache.FileLockManager.LockMode.OnDemand;
+import static org.gradle.internal.snapshot.CaseSensitivity.CASE_SENSITIVE;
+
+@SuppressWarnings("MethodMayBeStatic")
+class BuildCacheClientModule extends AbstractModule {
```

Comment:
🤔 I see we still have to initialize a lot of services. What is the idea with this client for future? 
I see that we currently just implement required types and wire everything together. Do we plan to also proof that it works?

Reply:
(We talked about this face-to-face as well, so just FTR:) The plan is to integrate as is for now, and later we can try to find ways to simplify creating some of these services.

As for testing, the idea is to turn the example client application into a "unit" test (not an integration test to avoid unwanted dependencies) to make sure we don't accidentally break these "APIs."